### ISSUE TITLE
intermittent "uninitialized constant ActiveStorage::Attached (NameError)" errors on Rails 6.1.6.1

### ISSUE THREAD

**Issue by thekendalmiller** [rails/rails#45600]
### Steps to reproduce
Unfortunately it is very random and I have not figured out how to reproduce it.  It happens when starting the environment. My guess is it's related to zeitwerk as we recently upgrades to Rails 6.x

```
irb(main):001:0> Rails.application.config.autoloader
=> :zeitwerk
irb(main):002:0> Rails.application.config.eager_load
=> true
```

The missing constant we've seen the most is `ActiveStorage::Attached`. Our application code does not reference that class directly, but does use ActiveStorage.

Here is the stack trace. 

```
/home/my_app/vendor/bundle/ruby/2.7.0/gems/activestorage-6.1.6.1/lib/active_storage/attached.rb:8:in `<module:ActiveStorage>': uninitialized constant ActiveStorage::Attached (NameError)

	from /home/my_app/vendor/bundle/ruby/2.7.0/gems/activestorage-6.1.6.1/lib/active_storage/attached.rb:5:in `<main>'
	from /home/my_app/vendor/bundle/ruby/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from /home/my_app/vendor/bundle/ruby/2.7.0/gems/bootsnap-1.12.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
	from /home/my_app/vendor/bundle/ruby/2.7.0/gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:35:in `require'
	from /home/my_app/vendor/bundle/ruby/2.7.0/gems/activestorage-6.1.6.1/lib/active_storage/engine.rb:118:in `block in <class:Engine>'
	from /home/my_app/vendor/bundle/ruby/2.7.0/gems/railties-6.1.6.1/lib/rails/initializable.rb:32:in `instance_exec'
	from /home/my_app/vendor/bundle/ruby/2.7.0/gems/railties-6.1.6.1/lib/rails/initializable.rb:32:in `run'
```

### Expected behavior
Rails application should start

### Actual behavior
Roughly 2% of the time, the application fails to start with a missing constant related to a rails class. When looking at it today, it's been `ActiveStorage::Attached`

### System configuration
**Rails version**: 6.1.6.1

**Ruby version**: 2.7.6p219

**Zeitwerk version**: 2.6.0

**Bootsnap version**: 1.12.0

**Comment by thekendalmiller** [comment]
It's also happening rrandomly with CI where `eager_load = false`.

**Comment by hahmed** [comment]
The only thing I can thing off is are you loading an ActiveRecord model somewhere in the initializers? Would be awesome if you could create a sample application that reproduces this, I could not get replicate the issue.

**Comment by thekendalmiller** [comment]
I haven't figured out how to make it happen. It hasn't happened since I opened the issue. There are no model references in the initializers. If I figure out how to make it happen consistently, I'll update this issue.

**Comment by natematykiewicz** [comment]
Anytime I've seen weird problems like this, Pry is installed. Zeitwerk and Pry don't play well together. Admittedly, usually the problem occurs when in a Pry debugging session and a constant is supposed to be autoloaded. I'm not sure if I've seen this Pry problem on app boot.

**Comment by thekendalmiller** [comment]
It hasn't happened since I opened this issue. I'll close it.
It started when we bumped rails from 6.1.6 to 6.1.6.1 and happened sporadically for 30 or so hours.  I did review that diff and nothing jumped out as what would cause this. It feels like a caching issue but I'm not sure how to reproduce it.

**Comment by maximevaillancourt** [comment]
@thekendalmiller, out of curiosity: were you seeing this issue only on GNU/Linux-based machines? (in other words, is that something you've seen happening on macOS/darwin-based machines?)

Do you have any new relevant information about this since the issue was closed?

For context: I'm facing the exact same problem, specifically in CI (GitHub Actions, Intel-based workers running Ubuntu 22.04), and I can't replicate this locally on a macOS (M1-based) machine. It happens randomly/intermittently/occasionally, making it very difficult (impossible so far for me) to replicate it reliably.

**Comment by thekendalmiller** [comment]
@maximevaillancourt We are seeing it on Ubuntu 22.04 and Amazon Linux 2(rhel/centos based). I've not been able to create a reproduction either - it's VERY VERY intermittent. This error happens maybe 1 out of 400 attempts to start rails server, rails console, or rails tests. 

My observation is it seems to happen shortly after bumping the `rails` gem version for minor releases. It's so infrequent for us that we just live with it. Running the command again without any other changes and it will then succeed.

It's not always the same constant that is missing, but it's been a rails constant each time, but that could just be coincidence.

**Comment by maximevaillancourt** [comment]
Interesting, thank you for sharing.

For the record: I've been seeing this issue with non-Rails constants too, such as other gems that we use within our Rails app _and_ internal constants as well.

Here are the constants we see missing every once in a while, even though we don't explicitly refer to all of them in our application code:
- `ActiveStorage::Attached`
- `ActiveStorage::Attached::Changes::CreateOne`
- `ActionView::TemplatePath`
- `Formtastic::Helpers::FormHelper`
- `RailsSemanticLogger::ActiveJob::LogSubscriber`
- `ViewComponent::Compiler`
- `ViewComponent::TestHelpers`

We've tried adding explicit `require` calls for these, but no dice, still seeing `uninitialized constant` exceptions pop up every now and then. I've tried removing/disabling pry/byebug based on a previous comment in this thread, but that didn't help either.

These seem to happen regardless of `eager_load` being `true` or `false`.

My current setup, in case it helps anyone: Rails 7.0.4 (Zeitwerk 2.6.6), Ruby 3.2.0, and Bootsnap 1.15.0.

**Comment by maximevaillancourt** [comment]
@fxn, this might be a long shot as I'm not confident this is related to Zeitwerk itself, but I've seen your name in multiple StackOverflow answers, GitHub issues, and Ruby bug tracker posts related to autoload/require and other loading problems, so I'm reaching out for help.

What would be the best way for me to help you help us? Any specific logs (collected during the flaky runs we've observed) we could share with you?

**Comment by maximevaillancourt** [comment]
@fxn / @byroot: this might be a long shot as I'm not confident this is related to Zeitwerk itself, but I've seen your names in multiple StackOverflow answers, GitHub issues, and Ruby bug tracker posts related to autoload/require and other loading problems, so I'm reaching out for help.

What would be the best way for me to help you help us? Any specific logs (collected during the flaky runs we've observed) we could share with you?

**Comment by maximevaillancourt** [comment]
As far as I can tell, all of the constants we've seen `NameError: uninitialized constant` exceptions for are related to gems implemented as engines (e.g. ActiveAdmin, a Rails engine [with some `autoload` calls](https://github.com/activeadmin/activeadmin/blob/f82cffcb2163baffff94057a25392630b232c15a/lib/active_admin.rb#L16-L55)), with the exception of our internal constants, which are not Rails engines and are not loaded through explicit `require` calls (they're just regular Ruby classes).

**Comment by fxn** [comment]
> which are not Rails engines and are not loaded through explicit require calls (they're just regular Ruby classes I'd expect to be autoloaded as usual).

I believe this is a good hint to try to follow, because if your regular internal gems occasionally miss constants, that suggests something unrelated to Rails or its autoloaders. Might be something more generic.

I'd like to understand in what sense do those gems not use `require`. In you open one, and assuming it consists of several files, how does it load its own code? And how does client clode load the gem?

**Comment by maximevaillancourt** [comment]
I think I miscommunicated what I meant to say, let me try again. Thank you for the help by the way, I appreciate it.

By "our internal constants", I mean regular class definitions, not even a gem. Here's one of the random/intermittent rspec failure messages we see for one of our internal constants (which just points to a plain old Ruby class):

```
An error occurred while loading ./spec/x/y/z/foo_spec.rb.
Failure/Error:
  RSpec.describe Z::FooForm do
    it "does the expected thing" do
      [truncated]

NameError:
  uninitialized constant Z::FooForm
```

`app/x/y/z/foo.rb` looks like this:

```rb
class Z::Foo < PlainOldParentClass
  # a regular looking class
end
```

Oddly, only this class leads into the `NameError` exception in CI intermittently. We have plenty of similar files in `app/x/y` (and its subdirectories) that work reliably 100% of the time.

---

Now, for the constants from external gems (e.g. `rails_semantic_logger`), these are required implicitly through the Gemfile (i.e. there's no explicit `require` call outside `Gemfile` for these, and there's no `require: false` for these in the Gemfile). As an additional example, we don't even have `formtastic` in our Gemfile — it's part of the `activeadmin` gem, which we do have in our Gemfile — yet we sometimes see `NameError` exceptions related it. See the following example exception that raises occasionally when running tests via rspec:

<details>
<summary>Example intermittent <code>NameError</code> exception for <code>Formtastic</code> when running specs in our CI flow (GitHub Actions)</summary>

```
Run ./bin/ci

An error occurred while loading rails_helper.
Failure/Error: require File.expand_path("../config/environment", __dir__)

NameError:
  uninitialized constant Formtastic::Helpers::FormHelper
# ./vendor/bundle/ruby/3.2.0/gems/formtastic-4.0.0/lib/formtastic/helpers/form_helper.rb:49:in `<module:Helpers>'
# ./vendor/bundle/ruby/3.2.0/gems/formtastic-4.0.0/lib/formtastic/helpers/form_helper.rb:2:in `<module:Formtastic>'
# ./vendor/bundle/ruby/3.2.0/gems/formtastic-4.0.0/lib/formtastic/helpers/form_helper.rb:1:in `<top (required)>'
# ./vendor/bundle/ruby/3.2.0/gems/bootsnap-1.15.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
# ./vendor/bundle/ruby/3.2.0/gems/bootsnap-1.15.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
# ./vendor/bundle/ruby/3.2.0/gems/zeitwerk-2.6.6/lib/zeitwerk/kernel.rb:38:in `require'
# ./vendor/bundle/ruby/3.2.0/gems/formtastic-4.0.0/lib/formtastic/engine.rb:7:in `block (2 levels) in <class:Engine>'
# ./vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4/lib/active_support/lazy_load_hooks.rb:95:in `class_eval'
# ./vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4/lib/active_support/lazy_load_hooks.rb:95:in `block in execute_hook'
# ./vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4/lib/active_support/lazy_load_hooks.rb:85:in `with_execution_control'
# ./vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4/lib/active_support/lazy_load_hooks.rb:90:in `execute_hook'
# ./vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4/lib/active_support/lazy_load_hooks.rb:60:in `block in on_load'
# ./vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4/lib/active_support/lazy_load_hooks.rb:59:in `each'
# ./vendor/bundle/ruby/3.2.0/gems/activesupport-7.0.4/lib/active_support/lazy_load_hooks.rb:59:in `on_load'
# ./vendor/bundle/ruby/3.2.0/gems/formtastic-4.0.0/lib/formtastic/engine.rb:6:in `block in <class:Engine>'
# ./vendor/bundle/ruby/3.2.0/gems/railties-7.0.4/lib/rails/initializable.rb:32:in `instance_exec'
# ./vendor/bundle/ruby/3.2.0/gems/railties-7.0.4/lib/rails/initializable.rb:32:in `run'
# ./vendor/bundle/ruby/3.2.0/gems/railties-7.0.4/lib/rails/initializable.rb:61:in `block in run_initializers'
# ./vendor/bundle/ruby/3.2.0/gems/railties-7.0.4/lib/rails/initializable.rb:60:in `run_initializers'
# ./vendor/bundle/ruby/3.2.0/gems/railties-7.0.4/lib/rails/application.rb:372:in `initialize!'
# ./config/environment.rb:7:in `<top (required)>'
# ./spec/rails_helper.rb:5:in `require'
# ./spec/rails_helper.rb:5:in `<top (required)>'
No examples found.
No examples found.


Finished in 0.00008 seconds (files took 3.57 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples

Finished in 0.00008 seconds (files took 3.57 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples
```
</details>

**Comment by maximevaillancourt** [comment]
I just noticed while looking at the final code snippet in my previous comment that rspec prints the last lines twice. Not sure if that's related to the rspec formatters, or if the specs ran twice somehow (which I guess could explain the `uninitialized constant` exception if constants got reloaded and the original one was unloaded?). I'll pull on this thread, it may be that we're somehow loading code twice in specs.

**Comment by fxn** [comment]
OK, let's move to that other ticket.

(BTW, `app/x/y/z/foo.rb` should normally define `Y::Z::Foo`, not `Z::Foo`. I presume things are correct and it was either a typo or the application has custom configuration that removes the `Y`.)

**Comment by maximevaillancourt** [comment]
> (BTW, app/x/y/z/foo.rb should normally define Y::Z::Foo, not Z::Foo. I presume things are correct and it was either a typo or the application has custom configuration that removes the Y.)

Right, this one's a bit weird because it's related to ActiveAdmin (`app/x` is actually `app/admin`), which [seems to handle `app/admin` in its own special way](https://github.com/activeadmin/activeadmin/blob/51689e48ec3faa51127b8462b2219e5e0c09ad8a/spec/unit/application_spec.rb#L8-L18).

**Comment by maximevaillancourt** [comment]
@thekendalmiller, do you use the `semantic_logger` or `rails_semantic_logger` gems in your project by any chance?

**Comment by thekendalmiller** [comment]
@maximevaillancourt We do use both of those gems. I'll catch up on your investigation. Thanks for following up!

**Comment by plicjo** [comment]
I have this resolved when upgrading to Ruby 3.3.6.

There seems to have been a bug fix regarding Bootsnap here: 

https://github.com/ruby/ruby/releases/tag/v3_3_2
https://bugs.ruby-lang.org/issues/20450
